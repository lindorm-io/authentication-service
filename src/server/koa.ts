import { CLIENT_CACHE_WORKER_OPTIONS, SERVER_PORT, TOKEN_ISSUER_MW_OPTIONS } from "../config";
import { KoaApp } from "@lindorm-io/koa";
import { NODE_ENVIRONMENT } from "../config";
import { NodeEnvironment } from "@lindorm-io/core";
import {
  appRoute,
  accountRoute,
  clientRoute,
  keyPairRoute,
  mfaRoute,
  oauthRoute,
  sessionRoute,
  userInfoRoute,
  wellKnownRoute,
} from "../route";
import { clientCacheWorker } from "@lindorm-io/koa-client";
import { keyPairCacheWorker, sessionCleanupWorker } from "../worker";
import { tokenIssuerMiddleware } from "@lindorm-io/koa-jwt";
import { winston } from "../logger";
import {
  cacheMiddleware,
  getMongoMiddleware,
  getRedisMiddleware,
  keystoreMiddleware,
  repositoryMiddleware,
} from "../middleware";

export const koa = new KoaApp({
  logger: winston,
  port: SERVER_PORT,
});

koa.addMiddleware(getMongoMiddleware());
koa.addMiddleware(repositoryMiddleware);
koa.addMiddleware(getRedisMiddleware());
koa.addMiddleware(cacheMiddleware);
koa.addMiddleware(keystoreMiddleware);
koa.addMiddleware(tokenIssuerMiddleware(TOKEN_ISSUER_MW_OPTIONS));

koa.addRoute("/", appRoute);
koa.addRoute("/account", accountRoute);
koa.addRoute("/client", clientRoute);
koa.addRoute("/key-pair", keyPairRoute);
koa.addRoute("/mfa", mfaRoute);
koa.addRoute("/oauth", oauthRoute);
koa.addRoute("/session", sessionRoute);
koa.addRoute("/userinfo", userInfoRoute);
koa.addRoute("/.well-known", wellKnownRoute);

if (NODE_ENVIRONMENT !== NodeEnvironment.TEST) {
  koa.addWorker(
    clientCacheWorker({
      ...CLIENT_CACHE_WORKER_OPTIONS,
      logger: winston,
    }),
  );
  koa.addWorker(keyPairCacheWorker);
  koa.addWorker(sessionCleanupWorker);
}
