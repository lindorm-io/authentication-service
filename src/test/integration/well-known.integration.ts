import MockDate from "mockdate";
import request from "supertest";
import { koa } from "../../server/koa";
import { v4 as uuid } from "uuid";
import { loadMongoConnection, loadRedisConnection, TEST_KEY_PAIR_REPOSITORY } from "../grey-box";

MockDate.set("2020-01-01 08:00:00.000");

describe("/.well-known", () => {
  beforeAll(async () => {
    await loadMongoConnection();
    await loadRedisConnection();
    koa.load();
  });

  test("GET /openid-configuration", async () => {
    const response = await request(koa.callback())
      .get(`/.well-known/openid-configuration`)
      .set("X-Correlation-ID", uuid())
      .expect(200);

    expect(response.body).toMatchSnapshot();
  });

  test("GET /jwks", async () => {
    const response = await request(koa.callback()).get(`/.well-known/jwks`).set("X-Correlation-ID", uuid()).expect(200);

    const [keyPair] = await TEST_KEY_PAIR_REPOSITORY.findMany({});

    expect(response.body).toStrictEqual({
      keys: [
        {
          alg: "ES512",
          c: 1577862000,
          e: "AQAB",
          exp: null,
          kid: keyPair.id,
          kty: "ec",
          n: keyPair.publicKey,
          use: "sig",
        },
      ],
    });
  });
});
