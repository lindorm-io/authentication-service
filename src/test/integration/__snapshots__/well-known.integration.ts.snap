// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/.well-known GET /openid-configuration 1`] = `
Object {
  "authorization_endpoint": "https://test.authentication.lindorm.io/oauth/authorization",
  "claims_parameter_supported": false,
  "grant_types_supported": Array [
    "pin_code",
    "biometrics",
    "email_link",
    "mfa-otp",
    "password",
    "refresh_token",
  ],
  "id_token_encryption_alg_values_supported": Array [],
  "id_token_encryption_enc_values_supported": Array [],
  "id_token_signing_alg_values_supported": Array [
    "ES512",
    "RS512",
  ],
  "issuer": "https://test.authentication.lindorm.io",
  "jwks_uri": "https://test.authentication.lindorm.io/.well-known/jwks",
  "request_parameter_supported": false,
  "request_uri_parameter_supported": false,
  "response_types_supported": Array [
    "token",
    "id_token",
    "code",
  ],
  "scopes_supported": Array [
    "default",
    "edit",
    "openid",
  ],
  "subject_types_supported": Array [],
  "token_endpoint": "https://test.authentication.lindorm.io/oauth/token",
  "token_endpoint_auth_methods_supported": Array [
    "biometrics",
    "email",
    "pin",
    "pwd",
    "token",
  ],
  "token_endpoint_auth_signing_alg_values_supported": Array [
    "ES512",
    "RS512",
  ],
  "userinfo_endpoint": "https://test.authentication.lindorm.io/userinfo",
}
`;
